File: workers/scraper/src/outreach_intelligence/routes/discovery.py
"""Lead discovery API routes."""

import logging
from typing import Any, Optional
from uuid import UUID

from fastapi import APIRouter, HTTPException
from pydantic import BaseModel, Field

from ..enrichment import EnrichmentAggregator
from ..scoring import ScoringEngine, SignalDetector
from ..models.icp import ICPProfile
from ..models.discovery import DiscoveredLead, DiscoveryStatus
from ..db.connection import fetch_one, fetch_all, execute_query

logger = logging.getLogger(__name__)
router = APIRouter(prefix="/api/v1/discover", tags=["discovery"])


class DiscoverLeadsRequest(BaseModel):
    """Request to discover new leads matching ICP."""

    user_id: UUID
    icp_id: Optional[UUID] = Field(
        default=None,
        description="ICP profile to use (uses default if not provided)",
    )
    limit: int = Field(default=25, ge=1, le=100)
    min_score: int = Field(
        default=60,
        ge=0,
        le=100,
        description="Minimum score threshold",
    )
    sources: Optional[list[str]] = Field(
        default=None,
        description="Sources to search (apollo, crunchbase)",
    )

    # API keys
    apollo_key: Optional[str] = None
    crunchbase_key: Optional[str] = None


class DiscoverLeadsResponse(BaseModel):
    """Response with discovered leads."""

    success: bool
    discovered: int
    above_threshold: int
    leads: list[dict[str, Any]]


class ReviewLeadRequest(BaseModel):
    """Request to review a discovered lead."""

    lead_id: UUID
    user_id: UUID
    action: str = Field(
        ...,
        description="Action to take: accept, reject, skip",
    )
    rejection_reason: Optional[str] = None


class SignalSearchRequest(BaseModel):
    """Request to search for signal-based leads."""

    user_id: UUID
    signal_type: str = Field(
        ...,
        description="Signal type: funding, hiring, tech_change",
    )
    days: int = Field(default=30, ge=1, le=90)
    industries: Optional[list[str]] = None
    min_funding: Optional[int] = None
    limit: int = Field(default=50, ge=1, le=100)

    crunchbase_key: Optional[str] = None
    apollo_key: Optional[str] = None


@router.post("", response_model=DiscoverLeadsResponse)
async def discover_leads(request: DiscoverLeadsRequest):
    """Discover new leads matching user's ICP.

    Searches across configured data sources for companies
    that match the ideal customer profile, then scores and
    ranks them.
    """
    try:
        # Load ICP
        icp = None
        icp_row = None

        if request.icp_id:
            icp_row = await fetch_one(
                "SELECT * FROM icp_profiles WHERE id = $1 AND user_id = $2",
                request.icp_id,
                request.user_id,
            )
        else:
            icp_row = await fetch_one(
                "SELECT * FROM icp_profiles WHERE user_id = $1 AND is_default = true",
                request.user_id,
            )

        if icp_row:
            icp = ICPProfile.from_db_row(dict(icp_row))

        if not icp:
            raise HTTPException(
                status_code=400,
                detail="No ICP profile found. Please create an ICP first.",
            )

        # Create aggregator
        aggregator = EnrichmentAggregator(
            user_id=request.user_id,
            apollo_key=request.apollo_key,
            crunchbase_key=request.crunchbase_key,
        )

        if not aggregator.apollo:
            raise HTTPException(
                status_code=400,
                detail="Apollo API key required for lead discovery.",
            )

        # Search for companies matching ICP
        search_result = await aggregator.apollo.search_companies(
            industries=[str(i) for i in icp.filters.industries] if icp.filters.industries else None,
            employee_count_min=icp.filters.company_size_min,
            employee_count_max=icp.filters.company_size_max,
            countries=[str(c) for c in icp.filters.countries] if icp.filters.countries else None,
            limit=request.limit * 2,  # Get extra to filter
        )

        if not search_result.success:
            raise HTTPException(
                status_code=500,
                detail=f"Search failed: {search_result.error}",
            )

        companies = search_result.data.get("companies", [])

        # Score and filter
        engine = ScoringEngine(icp=icp)
        discovered_leads = []

        for company in companies:
            # Check for duplicates
            existing = await fetch_one(
                """
                SELECT id FROM leads WHERE user_id = $1 AND company_domain = $2
                UNION
                SELECT id FROM discovered_leads WHERE user_id = $1 AND company_domain = $2
                """,
                request.user_id,
                company.get("domain"),
            )

            if existing:
                continue

            # Detect signals
            signals = SignalDetector.detect_signals(company)

            # Calculate preliminary score
            import uuid
            temp_lead_id = uuid.uuid4()

            score = await engine.score_lead(
                lead_id=temp_lead_id,
                company_data=company,
                signals=signals,
            )

            if score.total_score >= request.min_score:
                # Save to discovered_leads table
                discovered = DiscoveredLead(
                    user_id=request.user_id,
                    icp_id=icp.id,
                    company_name=company.get("name", "Unknown"),
                    company_domain=company.get("domain"),
                    company_linkedin=company.get("linkedin_url"),
                    company_data=company,
                    preliminary_score=score.total_score,
                    score_breakdown=score.score_breakdown.model_dump(),
                    discovery_signals=[s.model_dump() for s in signals],
                    source="apollo",
                )

                await _save_discovered_lead(discovered)

                discovered_leads.append({
                    "company_name": discovered.company_name,
                    "company_domain": discovered.company_domain,
                    "industry": company.get("industry"),
                    "employee_count": company.get("employee_count"),
                    "funding_stage": company.get("funding_stage"),
                    "score": score.total_score,
                    "tier": score.tier.value,
                    "signals": [s.signal_type.value for s in signals],
                })

                if len(discovered_leads) >= request.limit:
                    break

        await aggregator.close()

        return DiscoverLeadsResponse(
            success=True,
            discovered=len(discovered_leads),
            above_threshold=len(discovered_leads),
            leads=discovered_leads,
        )

    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"Lead discovery failed: {e}")
        raise HTTPException(status_code=500, detail=str(e))


@router.post("/signals")
async def discover_by_signals(request: SignalSearchRequest):
    """Discover leads based on specific signals.

    Find companies showing high-intent signals like:
    - Recent funding rounds
    - Active hiring
    - Technology changes
    """
    try:
        if request.signal_type == "funding":
            if not request.crunchbase_key:
                raise HTTPException(
                    status_code=400,
                    detail="Crunchbase API key required for funding signals.",
                )

            aggregator = EnrichmentAggregator(
                user_id=request.user_id,
                crunchbase_key=request.crunchbase_key,
            )

            companies = await aggregator.get_funding_signals(
                days=request.days,
                industries=request.industries,
                limit=request.limit,
            )

            await aggregator.close()

            return {
                "success": True,
                "signal_type": "funding",
                "count": len(companies),
                "companies": companies,
            }

        elif request.signal_type == "hiring":
            if not request.apollo_key:
                raise HTTPException(
                    status_code=400,
                    detail="Apollo API key required for hiring signals.",
                )

            # Use Apollo to find companies actively hiring
            aggregator = EnrichmentAggregator(
                user_id=request.user_id,
                apollo_key=request.apollo_key,
            )

            # Search for companies with open positions
            result = await aggregator.apollo.search_companies(
                industries=request.industries,
                limit=request.limit,
            )

            if result.success:
                # Filter to only hiring companies
                hiring_companies = [
                    c for c in result.data.get("companies", [])
                    if c.get("is_hiring") and c.get("open_positions", 0) > 0
                ]

                await aggregator.close()

                return {
                    "success": True,
                    "signal_type": "hiring",
                    "count": len(hiring_companies),
                    "companies": hiring_companies,
                }

            await aggregator.close()
            return {
                "success": False,
                "error": result.error,
            }

        else:
            raise HTTPException(
                status_code=400,
                detail=f"Unknown signal type: {request.signal_type}",
            )

    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"Signal discovery failed: {e}")
        raise HTTPException(status_code=500, detail=str(e))


@router.get("/pending/{user_id}")
async def get_pending_discoveries(
    user_id: UUID,
    limit: int = 50,
    offset: int = 0,
):
    """Get discovered leads pending review."""
    try:
        rows = await fetch_all(
            """
            SELECT * FROM discovered_leads
            WHERE user_id = $1 AND status = 'new'
            ORDER BY preliminary_score DESC
            LIMIT $2 OFFSET $3
            """,
            user_id,
            limit,
            offset,
        )

        leads = []
        for row in rows:
            leads.append({
                "id": str(row["id"]),
                "company_name": row["company_name"],
                "company_domain": row["company_domain"],
                "contact_name": row["contact_name"],
                "contact_title": row["contact_title"],
                "score": row["preliminary_score"],
                "source": row["source"],
                "discovered_at": row["discovered_at"].isoformat(),
            })

        # Get total count
        count_row = await fetch_one(
            "SELECT COUNT(*) as count FROM discovered_leads WHERE user_id = $1 AND status = 'new'",
            user_id,
        )

        return {
            "leads": leads,
            "count": len(leads),
            "total": count_row["count"] if count_row else 0,
        }

    except Exception as e:
        logger.error(f"Failed to get pending discoveries: {e}")
        raise HTTPException(status_code=500, detail=str(e))


@router.post("/review")
async def review_discovered_lead(request: ReviewLeadRequest):
    """Review and act on a discovered lead.

    Actions:
    - accept: Convert to actual lead
    - reject: Mark as rejected with reason
    - skip: Mark as reviewed but no action
    """
    try:
        # Get the discovered lead
        row = await fetch_one(
            "SELECT * FROM discovered_leads WHERE id = $1 AND user_id = $2",
            request.lead_id,
            request.user_id,
        )

        if not row:
            raise HTTPException(status_code=404, detail="Discovered lead not found")

        if request.action == "accept":
            # Create actual lead from discovered lead
            lead_id = await _convert_to_lead(dict(row))

            # Update discovered lead status
            await execute_query(
                """
                UPDATE discovered_leads
                SET status = 'accepted', reviewed_at = NOW(), accepted_at = NOW(),
                    converted_lead_id = $2
                WHERE id = $1
                """,
                request.lead_id,
                lead_id,
            )

            return {
                "success": True,
                "action": "accepted",
                "lead_id": str(lead_id),
            }

        elif request.action == "reject":
            await execute_query(
                """
                UPDATE discovered_leads
                SET status = 'rejected', reviewed_at = NOW(), rejection_reason = $2
                WHERE id = $1
                """,
                request.lead_id,
                request.rejection_reason,
            )

            return {
                "success": True,
                "action": "rejected",
            }

        elif request.action == "skip":
            await execute_query(
                """
                UPDATE discovered_leads
                SET status = 'reviewed', reviewed_at = NOW()
                WHERE id = $1
                """,
                request.lead_id,
            )

            return {
                "success": True,
                "action": "skipped",
            }

        else:
            raise HTTPException(
                status_code=400,
                detail=f"Invalid action: {request.action}",
            )

    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"Failed to review lead: {e}")
        raise HTTPException(status_code=500, detail=str(e))


async def _save_discovered_lead(lead: DiscoveredLead) -> UUID:
    """Save discovered lead to database."""
    import json
    import uuid

    lead_id = uuid.uuid4()

    await execute_query(
        """
        INSERT INTO discovered_leads (
            id, user_id, icp_id, company_name, company_domain, company_linkedin,
            contact_name, contact_title, contact_email, contact_linkedin,
            company_data, contact_data, preliminary_score, score_breakdown,
            discovery_signals, status, source, discovered_at
        )
        VALUES ($1, $2, $3, $4, $5, $6, $7, $8, $9, $10, $11, $12, $13, $14, $15, $16, $17, NOW())
        """,
        lead_id,
        lead.user_id,
        lead.icp_id,
        lead.company_name,
        lead.company_domain,
        lead.company_linkedin,
        lead.contact_name,
        lead.contact_title,
        lead.contact_email,
        lead.contact_linkedin,
        json.dumps(lead.company_data),
        json.dumps(lead.contact_data),
        lead.preliminary_score,
        json.dumps(lead.score_breakdown),
        json.dumps(lead.discovery_signals),
        lead.status.value,
        lead.source,
    )

    return lead_id


async def _convert_to_lead(discovered: dict) -> UUID:
    """Convert discovered lead to actual lead."""
    import json
    import uuid

    lead_id = uuid.uuid4()

    await execute_query(
        """
        INSERT INTO leads (
            id, user_id, company_name, company_domain, name, title, email,
            linkedin_url, company_data, status, source, created_at
        )
        VALUES ($1, $2, $3, $4, $5, $6, $7, $8, $9, 'new', $10, NOW())
        """,
        lead_id,
        discovered["user_id"],
        discovered["company_name"],
        discovered["company_domain"],
        discovered["contact_name"],
        discovered["contact_title"],
        discovered["contact_email"],
        discovered["contact_linkedin"],
        json.dumps(discovered["company_data"]) if discovered["company_data"] else "{}",
        discovered["source"] or "discovery",
    )

    return lead_id

File: apps/api/src/routes/discovery.ts
/**
 * Lead Discovery Routes
 * Proxy to the Intelligence Worker for lead discovery and enrichment
 */

import { Router, Response } from 'express';
import { authMiddleware, AuthenticatedRequest } from '../middleware/auth.js';
import { config } from '../config.js';
import { query } from '../db.js';
import { validateUUID } from '../utils/validation.js';

const router = Router();

// All routes require authentication
router.use(authMiddleware);

// Intelligence worker URL
const INTELLIGENCE_URL = process.env.INTELLIGENCE_WORKER_URL || 'http://localhost:8000';

/**
 * POST /discovery/enrich-company
 * Enrich company data
 */
router.post('/enrich-company', async (req: AuthenticatedRequest, res: Response) => {
  try {
    const { domain, sources } = req.body;

    if (!domain) {
      return res.status(400).json({ error: 'Domain is required' });
    }

    // Get user's API credentials
    const credentials = await getUserCredentials(req.userId!);

    // Call intelligence worker
    const response = await fetch(`${INTELLIGENCE_URL}/api/v1/enrich/company`, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({
        domain,
        sources,
        user_id: req.userId,
        ...credentials,
      }),
    });

    const result = await response.json();
    res.status(response.status).json(result);
  } catch (error) {
    console.error('Enrichment failed:', error);
    res.status(500).json({ error: 'Enrichment service unavailable' });
  }
});

/**
 * POST /discovery/find-contacts
 * Find contacts at a company
 */
router.post('/find-contacts', async (req: AuthenticatedRequest, res: Response) => {
  try {
    const { domain, titles, departments, limit = 5 } = req.body;

    if (!domain) {
      return res.status(400).json({ error: 'Domain is required' });
    }

    const credentials = await getUserCredentials(req.userId!);

    const response = await fetch(`${INTELLIGENCE_URL}/api/v1/enrich/contacts`, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({
        domain,
        titles,
        departments,
        limit,
        user_id: req.userId,
        ...credentials,
      }),
    });

    const result = await response.json();
    res.status(response.status).json(result);
  } catch (error) {
    console.error('Contact finding failed:', error);
    res.status(500).json({ error: 'Contact finding service unavailable' });
  }
});

/**
 * POST /discovery/verify-email
 * Verify an email address
 */
router.post('/verify-email', async (req: AuthenticatedRequest, res: Response) => {
  try {
    const { email } = req.body;

    if (!email) {
      return res.status(400).json({ error: 'Email is required' });
    }

    const credentials = await getUserCredentials(req.userId!);

    const response = await fetch(`${INTELLIGENCE_URL}/api/v1/enrich/verify-email`, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({
        email,
        user_id: req.userId,
        hunter_key: credentials.hunter_key,
      }),
    });

    const result = await response.json();
    res.status(response.status).json(result);
  } catch (error) {
    console.error('Email verification failed:', error);
    res.status(500).json({ error: 'Email verification service unavailable' });
  }
});

/**
 * POST /discovery/discover
 * Discover new leads matching ICP
 */
router.post('/discover', async (req: AuthenticatedRequest, res: Response) => {
  try {
    const { icp_id, limit = 25, min_score = 60, sources } = req.body;

    const credentials = await getUserCredentials(req.userId!);

    const response = await fetch(`${INTELLIGENCE_URL}/api/v1/discover`, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({
        user_id: req.userId,
        icp_id,
        limit,
        min_score,
        sources,
        ...credentials,
      }),
    });

    const result = await response.json();
    res.status(response.status).json(result);
  } catch (error) {
    console.error('Discovery failed:', error);
    res.status(500).json({ error: 'Discovery service unavailable' });
  }
});

/**
 * GET /discovery/pending
 * Get discovered leads pending review
 */
router.get('/pending', async (req: AuthenticatedRequest, res: Response) => {
  try {
    const limit = parseInt(req.query.limit as string) || 50;
    const offset = parseInt(req.query.offset as string) || 0;

    const result = await query(
      `SELECT * FROM discovered_leads
       WHERE user_id = $1 AND status = 'new'
       ORDER BY preliminary_score DESC
       LIMIT $2 OFFSET $3`,
      [req.userId, limit, offset]
    );

    const countResult = await query(
      `SELECT COUNT(*) as count FROM discovered_leads
       WHERE user_id = $1 AND status = 'new'`,
      [req.userId]
    );

    res.json({
      leads: result.rows,
      count: result.rows.length,
      total: parseInt(countResult.rows[0].count),
    });
  } catch (error) {
    console.error('Failed to get pending discoveries:', error);
    res.status(500).json({ error: 'Failed to get pending discoveries' });
  }
});

/**
 * POST /discovery/:id/review
 * Review a discovered lead (accept/reject)
 */
router.post('/:id/review', async (req: AuthenticatedRequest, res: Response) => {
  try {
    const { id } = req.params;
    const { action, rejection_reason } = req.body;

    if (!validateUUID(id)) {
      return res.status(400).json({ error: 'Invalid lead ID' });
    }

    if (!['accept', 'reject', 'skip'].includes(action)) {
      return res.status(400).json({ error: 'Invalid action' });
    }

    // Get the discovered lead
    const leadResult = await query(
      `SELECT * FROM discovered_leads WHERE id = $1 AND user_id = $2`,
      [id, req.userId]
    );

    if (leadResult.rows.length === 0) {
      return res.status(404).json({ error: 'Discovered lead not found' });
    }

    const discovered = leadResult.rows[0];

    if (action === 'accept') {
      // Create actual lead
      const newLeadResult = await query(
        `INSERT INTO leads (
          user_id, company_name, company_domain, name, title, email,
          linkedin_url, company_data, status, source
        )
        VALUES ($1, $2, $3, $4, $5, $6, $7, $8, 'new', 'discovery')
        RETURNING id`,
        [
          req.userId,
          discovered.company_name,
          discovered.company_domain,
          discovered.contact_name,
          discovered.contact_title,
          discovered.contact_email,
          discovered.contact_linkedin,
          discovered.company_data,
        ]
      );

      // Update discovered lead status
      await query(
        `UPDATE discovered_leads
         SET status = 'accepted', reviewed_at = NOW(), accepted_at = NOW(),
             converted_lead_id = $2
         WHERE id = $1`,
        [id, newLeadResult.rows[0].id]
      );

      res.json({
        success: true,
        action: 'accepted',
        lead_id: newLeadResult.rows[0].id,
      });
    } else if (action === 'reject') {
      await query(
        `UPDATE discovered_leads
         SET status = 'rejected', reviewed_at = NOW(), rejection_reason = $2
         WHERE id = $1`,
        [id, rejection_reason]
      );

      res.json({ success: true, action: 'rejected' });
    } else {
      await query(
        `UPDATE discovered_leads
         SET status = 'reviewed', reviewed_at = NOW()
         WHERE id = $1`,
        [id]
      );

      res.json({ success: true, action: 'skipped' });
    }
  } catch (error) {
    console.error('Failed to review lead:', error);
    res.status(500).json({ error: 'Failed to review lead' });
  }
});

/**
 * GET /discovery/scores
 * Get lead score distribution
 */
router.get('/scores', async (req: AuthenticatedRequest, res: Response) => {
  try {
    const result = await query(
      `SELECT tier, COUNT(*) as count, AVG(total_score) as avg_score
       FROM lead_current_scores lcs
       JOIN leads l ON lcs.lead_id = l.id
       WHERE l.user_id = $1
       GROUP BY tier
       ORDER BY avg_score DESC`,
      [req.userId]
    );

    const distribution = {
      hot: { count: 0, avg_score: 0 },
      warm: { count: 0, avg_score: 0 },
      nurture: { count: 0, avg_score: 0 },
      cold: { count: 0, avg_score: 0 },
    };

    for (const row of result.rows) {
      if (row.tier in distribution) {
        distribution[row.tier as keyof typeof distribution] = {
          count: parseInt(row.count),
          avg_score: Math.round(parseFloat(row.avg_score) * 10) / 10,
        };
      }
    }

    const total = Object.values(distribution).reduce((sum, d) => sum + d.count, 0);

    res.json({
      total_leads: total,
      distribution,
    });
  } catch (error) {
    console.error('Failed to get score distribution:', error);
    res.status(500).json({ error: 'Failed to get score distribution' });
  }
});

/**
 * GET /discovery/scored-leads
 * Get scored leads
 */
router.get('/scored-leads', async (req: AuthenticatedRequest, res: Response) => {
  try {
    const tier = req.query.tier as string;
    const minScore = parseInt(req.query.min_score as string) || 0;
    const limit = parseInt(req.query.limit as string) || 50;
    const offset = parseInt(req.query.offset as string) || 0;

    let whereClause = 'WHERE l.user_id = $1';
    const params: any[] = [req.userId];

    if (tier) {
      params.push(tier);
      whereClause += ` AND lcs.tier = $${params.length}`;
    }

    if (minScore > 0) {
      params.push(minScore);
      whereClause += ` AND lcs.total_score >= $${params.length}`;
    }

    params.push(limit, offset);

    const result = await query(
      `SELECT l.*, lcs.total_score, lcs.tier, lcs.intent_score,
              lcs.fit_score, lcs.accessibility_score, lcs.score_breakdown,
              lcs.calculated_at as scored_at
       FROM leads l
       JOIN lead_current_scores lcs ON l.id = lcs.lead_id
       ${whereClause}
       ORDER BY lcs.total_score DESC
       LIMIT $${params.length - 1} OFFSET $${params.length}`,
      params
    );

    res.json({
      leads: result.rows,
      count: result.rows.length,
    });
  } catch (error) {
    console.error('Failed to get scored leads:', error);
    res.status(500).json({ error: 'Failed to get scored leads' });
  }
});

/**
 * POST /discovery/credentials
 * Save user's enrichment API credentials
 */
router.post('/credentials', async (req: AuthenticatedRequest, res: Response) => {
  try {
    const { service, api_key } = req.body;

    const validServices = ['apollo', 'clearbit', 'hunter', 'crunchbase'];
    if (!validServices.includes(service)) {
      return res.status(400).json({ error: 'Invalid service' });
    }

    if (!api_key || api_key.trim().length === 0) {
      return res.status(400).json({ error: 'API key is required' });
    }

    // Store the key (in production, this should be encrypted)
    // For now, store as-is with a hint
    const keyHint = api_key.slice(-4);

    await query(
      `INSERT INTO enrichment_credentials (user_id, service, api_key_encrypted, api_key_hint, is_valid)
       VALUES ($1, $2, $3, $4, true)
       ON CONFLICT (user_id, service)
       DO UPDATE SET api_key_encrypted = $3, api_key_hint = $4, is_valid = true, updated_at = NOW()`,
      [req.userId, service, Buffer.from(api_key), keyHint]
    );

    res.json({
      success: true,
      service,
      hint: `****${keyHint}`,
    });
  } catch (error) {
    console.error('Failed to save credentials:', error);
    res.status(500).json({ error: 'Failed to save credentials' });
  }
});

/**
 * GET /discovery/credentials
 * Get user's configured credentials (hints only)
 */
router.get('/credentials', async (req: AuthenticatedRequest, res: Response) => {
  try {
    const result = await query(
      `SELECT service, api_key_hint, is_valid, last_validated_at, error_message,
              credits_remaining, credits_limit
       FROM enrichment_credentials
       WHERE user_id = $1`,
      [req.userId]
    );

    res.json({
      credentials: result.rows.map((row) => ({
        service: row.service,
        configured: true,
        hint: row.api_key_hint ? `****${row.api_key_hint}` : null,
        is_valid: row.is_valid,
        error: row.error_message,
        credits_remaining: row.credits_remaining,
        credits_limit: row.credits_limit,
      })),
    });
  } catch (error) {
    console.error('Failed to get credentials:', error);
    res.status(500).json({ error: 'Failed to get credentials' });
  }
});

/**
 * DELETE /discovery/credentials/:service
 * Remove a credential
 */
router.delete('/credentials/:service', async (req: AuthenticatedRequest, res: Response) => {
  try {
    const { service } = req.params;

    await query(
      `DELETE FROM enrichment_credentials WHERE user_id = $1 AND service = $2`,
      [req.userId, service]
    );

    res.json({ success: true });
  } catch (error) {
    console.error('Failed to delete credential:', error);
    res.status(500).json({ error: 'Failed to delete credential' });
  }
});

// Helper function to get user's API credentials
async function getUserCredentials(userId: string): Promise<Record<string, string>> {
  const result = await query(
    `SELECT service, api_key_encrypted FROM enrichment_credentials
     WHERE user_id = $1 AND is_valid = true`,
    [userId]
  );

  const credentials: Record<string, string> = {};

  for (const row of result.rows) {
    // In production, decrypt the key here
    const key = row.api_key_encrypted.toString();
    credentials[`${row.service}_key`] = key;
  }

  return credentials;
}

export default router;
